### PR TITLE
Average win lost points

### PR DESCRIPTION
**Pull request by alexeydergunov**

### CODE REVIEW

File: Mimir/src/models/PlayerStat.php [hunk 1]
```diff
@@ -200,6 +200,92 @@ protected function _getPlacesSummary(int $playerId, array $games)
...
+     * @param int $honba
+     * @param int $riichiBetsCount
+     * @param bool $isTsumo
+     * @param int|null $paoPlayerId
+     * @param int|null $closestWinnerForMultiron
+     * @param int $totalRiichiInRoundForMultiron
+     * @return int
+     * @throws InvalidParametersException
+     */
+    protected function _getPointsDelta(\Common\Ruleset $rules, array $currentScores, int $han, ?int $fu, int $winnerId, ?int $loserId, array $riichiIds, int $playerId, int $dealerId, int $honba, int $riichiBetsCount, bool $isTsumo, ?int $paoPlayerId, ?int $closestWinnerForMultiron = null, int $totalRiichiInRoundForMultiron = 0)
```
Comment (ctizen) [inline Mimir/src/models/PlayerStat.php:222]: Could you please remove type hints for now? In general, it's not guaranteed that input types would match expected ones, at least for now, and I don't think we want to ensure strict typing across all php codebase in pantheon. The bad thing is, when actual input type doesn't match the expected one, the whole application just fails with an error. So I think having types in phpdoc is more than enough now.

Comment (ctizen) [reply Mimir/src/models/PlayerStat.php:222]: Also I'd ask to separate args with line breaks. There's too many of them, so multiline list would be easier to read

File: Mimir/src/models/PlayerStat.php [hunk 2]
```diff
@@ -282,6 +384,7 @@ protected function _getOutcomeSummary(int $playerId, array $rounds)
                             break;
                         } else if ($round->getLoserId() == $playerId) {
                             $acc['feed'] ++;
+                            $acc['points_lost_ron'] -= $this->_getPointsDeltaForRound($round, $playerId, false, $multironRiichiWinners);
```
Comment (ctizen) [inline Mimir/src/models/PlayerStat.php:387]: Looks like the `_getPointsDeltaForRound` function is called twice with same arguments and it's quite heavy. I'd suggest memoizing it using some local static cache. Example: https://eddmann.com/posts/implementing-and-using-memoization-in-php/

File: Mimir/src/models/PlayerStat.php [hunk 3]
```diff
@@ -200,6 +200,92 @@ protected function _getPlacesSummary(int $playerId, array $games)
...
+            $closestWinnerForMultiron = null;
+            $totalRiichiInRoundForMultiron = 0;
+            $honba = $sessionState->getHonba();
+            $riichiBetsCount = $sessionState->getRiichiBets();
+        }
+
+        if ($winnerId != $playerId) {
+            // if we deal-in or someone else takes tsumo, don't consider any riichi sticks in the current round
+            // since we are interested only in deal-in cost
+            $riichiIds = array();
```
Comment (ctizen) [inline Mimir/src/models/PlayerStat.php:266]: There's shortcut that just works: `$riichiIds = []`

File: Mimir/src/models/PlayerStat.php (hunk 2 above)
Comment (alexeydergunov) [reply Mimir/src/models/PlayerStat.php:387]: Its calls are in separate code branches, so it is not called twice, but I can move the call before if

File: Mimir/src/models/PlayerStat.php (hunk 1 above)
Comment (alexeydergunov) [reply Mimir/src/models/PlayerStat.php]: Done

File: Mimir/src/models/PlayerStat.php (hunk 3 above)
Comment (alexeydergunov) [reply Mimir/src/models/PlayerStat.php]: Done

**Review by ctizen — APPROVED** [review]

**Pull request merged**